### DISCUSSION TITLE
How to get a payload size of TCP packets?

### DISCUSSION THREAD

**Discussion by vmprog** [mitmproxy/mitmproxy#5275] [Q&A] [answered]
Hi!
My goal: To get the payload size of TCP packets, that could not be proxied (when Client TLS handshake failed).

Something like:
response_body_decoded_size = len(flow.response.content) if flow.response.content is still 0

My train of thought is like this. I have to go through all the messages in the stream and get len(message.content) for each one.
```python
def tcp_message(stream: tcp.TCP stream):
message = stream.messages[-1]
ctx.log(f"tcp_message: {flow=}")                                            
ctx.log(f"len req: {len(message.content)}")
```
Is this the right approach?

**Comment by mhils** [comment, marked as answer]
> My goal: To get the payload size of TCP packets, that could not be proxied

What for? If your goal is to obtain the failed TLS handshakes, I would recommend using the [`tls_clienthello`](https://docs.mitmproxy.org/dev/api/events.html#TLSEvents.tls_clienthello) to store the handshake and then reuse it in one of the client/server established/failed hooks.

To answer your original question, mitmproxy does not really operate on the level of individual packets, we only work with socket objects. So obtaining the original payload sizes is not really possible. Even for TCP streams, you may see multiple packets end up in a single "Message", see https://docs.mitmproxy.org/dev/api/mitmproxy/tcp.html#TCPMessage.

**Reply by vmprog** [reply]
Thanks. It became clearer.
